File: app/src/main/java/com/example/ge63vr/presenter/Presenter.java
package com.example.ge63vr.presenter;

import com.example.ge63vr.model.Data;
import com.example.ge63vr.model.MVPContrat;
import com.example.ge63vr.model.Model;
import com.example.ge63vr.view.MainActivity;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class Presenter implements MVPContrat.Presenter {
    private MainActivity mainActivity;
    private Model mModel= new Model();

    public Presenter(MainActivity mainActivity) {
        this.mainActivity = mainActivity;
    }

    public void get(Call<Data> dataCall) {
        dataCall.enqueue(new Callback<Data>() {
            @Override
            public void onResponse(Call<Data> call, Response<Data> response) {
                mainActivity.onUpdate(response.body());
            }

            @Override
            public void onFailure(Call<Data> call, Throwable t) {
                mainActivity.onError();
            }
        });
    }

    @Override
    public void getNews() {
        get(mModel.getLatestNews());
    }

    @Override
    public void refresh() {
    mainActivity.clearData();
    mModel.getLatestNews();
    }

    @Override
    public void loadMore(String date) {
        get(mModel.getBeforeNews(date));

    }

}

File: app/src/main/java/com/example/ge63vr/presenter/Presenter.kt
package com.example.ge63vr.presenter


import android.os.Handler
import android.os.Looper
import com.example.ge63vr.model.Data
import com.example.ge63vr.model.MVPContrat
import com.example.ge63vr.model.Model
import com.example.ge63vr.view.MainActivity
import kotlinx.coroutines.experimental.android.HandlerContext
import kotlinx.coroutines.experimental.launch
import retrofit2.Call


class Presenter(private val mainActivity: MainActivity) : MVPContrat.Presenter {
    private val mModel = Model()
    val UI = HandlerContext(Handler(Looper.getMainLooper()), "UI")
    operator fun get(dataCall: Call<Data>?) {
        launch {
            val databean:Data =dataCall?.execute()!!.body()
            launch(UI) {
                mainActivity.onUpdate(databean)
            }
        }

        /*dataCall?.enqueue(object : Callback<Data> {
            override fun onResponse(call: Call<Data>, response: Response<Data>) {
                mainActivity.onUpdate(response.body())
            }

            override fun onFailure(call: Call<Data>, t: Throwable) {
                mainActivity.onError()
            }
        })*/
    }

    override fun getNews() {
        get(mModel.latestNews)
    }

    override fun refresh() {
        mainActivity.clearData()
        mModel.latestNews
    }

    override fun loadMore(date: String?) {
        get(mModel.getBeforeNews(date))
    }

}
